### DISCUSSION TITLE
How should I add backends?

### DISCUSSION THREAD

**Discussion by yuzumx** [company-mode/company-mode#1196] [Q&A] [answered]
Hello, and sorry for my poor English.

How to add backends to `company-backends` is more appropriate?

Is it necessary to make this variable into buffer-local?

Thanks.

**Comment by cyrialize** [comment, marked as answer]
There are three ways to do this - you can add backends globally, locally, or both. There's some pros and cons to each (anyone else - feel free to jump in and correct me if I am wrong). For these examples I'll use the backend, company-ac-php-backend, a php completion backend.

Adding backends to the global company-backends variable, ex:
```
(add-to-list 'company-backends 'company-ac-php-backend)
```
Requires the least amount of code, but then that means the backend you just added will be available anywhere that you have company-mode enabled. This means that you could be in a .txt file and if you start to type the beginning of a php function you'll get PHP completions as part of the output. In php there's the function ```isset()```, so there's a possibility that if you start to type is, then the completion for ```isset()``` will appear.

Adding just backends locally is great if you want backends to only be available for certain buffers while keeping company-mode enabled globally. Here's a snippet of my elisp code for php:
```
(use-package php-mode
  :mode
  (("[^.][^t][^p][^l]\\.php$" . php-mode))
  :config
  (add-hook 'php-mode-hook
'(lambda ()
     (set (make-local-variable 'company-backends)
            '((company-dabbrev-code company-ac-php-backend)))
```

What this does is that now instead of company using the global company-backends, it uses the locally defined company-backends while in php-mode buffers. This is great if you want specific backends to work for specific buffers, like a php backend only appearing for php code.

The third option, globally and locally is currently what I use. I have company-mode enable globally with backends (company-dabbrev, etc) part of the global-backends. I use org-mode, so having completion available is nice when I'm taking notes. I also use the snippet above for php-code, when I only want specific backends available. 

So to answer your question - it is not necessary to make company-backends buffer-local, but there are specific use cases in doing so.
